Free Listings + Paid Ads: Combine the audience and shipping steps

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

It's a part of **📌 Adjust the setting sections in the onboarding steps** in #1610. Mainly focus on combining the audience and shipping steps for the onboarding flow and the editing free listings page. Most visual and wording changes are not included.

- Move the `SetupFreeListings` component of editing free listings to the shared directory.
- Change the shared `ChooseAudience` to `ChooseAudienceSection` and merge it into `SetupFreeListings` with corresponding validations for the editing free listings.
- Replace `SetupFreeListings` with the shared one for the onboarding flow.
- Remove unused files and an npm package `use-debounce`.

It also handles some issues:

- There are no `useGetRemainingCountryCodes` hooks after this PR.
   - #274
- Components of shipping times are changed to use the shared ones. Each group of shipping rates and shipping times components is no longer duplicated.
   - #288
- The last shipping values flash issue only happens on the inputs of shipping time during onboarding setup. With the replacement of using shared shipping time components, the form state management was isolated from wp-store.
   - #1078
- After replacing `ChooseAudience` and `SetupFreeListings` with the shared components, the auto-saving mechanism is implemented with `on*Change` callbacks instead. All auto-saving hooks and debouncing are no longer in need.
   - #1335

### Screenshots:

#### 📷 Onboarding flow

![wp test_wp-admin_admin php_page=wc-admin path=%2Fgoogle%2Fsetup-mc google-mc=connected](https://user-images.githubusercontent.com/17420811/182823358-348f2654-9e1d-4c03-ba73-7874dfbb2294.png)

#### 📷 Edit free listings

![wp test_wp-admin_admin php_page=wc-admin path=%2Fgoogle%2Fproduct-feed google-mc=connected guide=submission-success (2)](https://user-images.githubusercontent.com/17420811/182827431-e70c87fe-7a79-4ae0-9599-a660ea3d15d1.png)

### Detailed test instructions:

#### Onboarding flow

1. Disconnect all accounts or run these SQLs to delete related GLA settings
   ```sql
   DELETE FROM wp_options WHERE option_name IN ('gla_mc_setup_completed_at', 'gla_target_audience', 'gla_merchant_center');
   DELETE FROM wp_gla_shipping_rates;
   DELETE FROM wp_gla_shipping_times;
   ```
2. Start the onboarding setup and check if the combined step for audience and shipping works correctly.
   - Auto-saving mechanism for audience, MC settings, shipping rates, and shipping times
   - All possible ways to add/edit shipping times
   - Switch the location options of audience to see if **all** countries and **selected** countries options affect the corresponding shipping sections correctly.
3. Change the audience countries to see if the removed countries are removed from shipping sections as well.
   - 💡 The implementation is not the same as before. Please refer to the originally expected behavior in https://github.com/woocommerce/google-listings-and-ads/issues/232#issuecomment-785687459
4. Edit settings to check if the disabled status of the "Continue" button is correct.
5. Continue or switch between steps to view whether it works as expected.
6. Refer to the replicate steps in #1078 to check if the value flash problem in shipping time fields is fixed.
   - 💡 The problem in shipping rate fields has been fixed in another PR before.

#### Edit free listings

1. Go to the Edit Free Listings page.
2. It should be changed to a one-pager without steps.
3. Simiar to the tests in the onboarding flow except for the auto-saving mechanism, operate the combined audience and shipping to see if they work well.
4. Edit some values to see if the unsaved data prompt pops up when navigating away.
5. Edit the above values back to see if there is no prompt when navigating away.

### Additional details:

- The auto-fill feature of flat shipping rates in the onboarding flow is removed from the frontend side. Ref: p1659494048432789-slack-C01E9LB4X61
- Not sure why but `mc/target_audience` will respond with an 400 error if saving with empty countries. From my memory, it should be ok before because the UI allows merchants to unselect all. 🤔 

### Changelog entry

> Add - Combine the audience and shipping steps for the onboarding flow and the editing free listings page.
> Fix - Shipping time values flash during the onboarding setup.

eason9487 committed:
Move SetupFreeListings component of editing free listings to the shared directory

eason9487 committed:
Change the shared ChooseAudience to ChooseAudienceSection and merge it into SetupFreeListings for the editing free listings page

eason9487 committed:
Add validations for merged audience values in the shared SetupFreeListings

eason9487 committed:
Replace SetupFreeListings with the shared one for the onboarding flow

eason9487 committed:
Catch errors when saving data in the SavedSetupStepper of onboarding flow

eason9487 committed:
Remove unused files

eason9487 committed:
Remove an unused npm package use-debounce

eason9487 committed:
Add the CountryCode type back to SetupFreeListings

eason9487 committed:
Add auto-saving for MC settings to fall back with the original implementation

puntope commented on js/src/components/free-listings/configure-product-listings/__snapshots__/checkErrors.test.js.snap:
```diff
@@ -1,5 +1,11 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
+exports[`checkErrors Audience When the audience countries array is empty and the value of audience location option is 'selected', should not pass 1`] = `"Please select at least one country."`;
```
💅  Out of this PR and opinionated

Maybe to consider migration to unit testing?

eason9487 replied:
Do you mean replacing `expect( errors.countries ).toBe( 'Please select at least one country.' );` with `expect( errors.countries ).toMatchSnapshot();` ? If so, some related discussions can be found in https://github.com/woocommerce/google-listings-and-ads/pull/893#discussion_r674040117 and https://github.com/woocommerce/google-listings-and-ads/pull/927#discussion_r681107540.

puntope commented on js/src/setup-mc/setup-stepper/saved-setup-stepper.js:
```diff
@@ -25,6 +33,49 @@ import stepNameKeyMap from './stepNameKeyMap';
+		if ( settings?.shipping_rate === null ) {
+			saveSettings( {
+				...settings,
+				shipping_rate: 'automatic',
+				shipping_time: 'flat',
```
❓ Not a suggestion, just curious. Why we are not adding the default values inside data store DEFAULT_STATE ?

puntope replied:
I saw we use null values as checkers for other stuff... probably because of that

eason9487 replied:
The intent here is to trigger API to save the form's default values. It would be easier and more suitable to handle the intent here.

In addition, moving the default values of form layer to wp-data store layer will also bring different abstraction of data/states isolation. Since `isResolving` and `hasFinishedResolution` are not enough to cover all statuses of a state in wp-store (see issue #438.), and considering the "statuses" of states in wp-data haven't yet had a consistent definition/convention, for now, the default value in a wp-data state is more like the status that tells the state is not ready to be used. And after receiving the data from API, the state will be changed to that data, which will also change the status of that state to "the status of the state is ready, and the data is the state itself".

IMO, before a consistent definition/convention of the statuses, states, and data in wp-data is created, it's probably hard to say how to set a default value would be a good practice.

puntope commented on js/src/setup-mc/setup-stepper/saved-setup-stepper.js:
```diff
@@ -58,6 +100,15 @@ const SavedSetupStepper = ( { savedStep, onRefetchSavedStep = () => {} } ) => {
 	};
 
+	function handleFormChange( errorMessage, newValue ) {
+		this( newValue ).catch( () => createNotice( 'error', errorMessage ) );
```
💅  it was a bit confusing to understand what `this` was doing here. I saw after  in the code that we bind  the action  to be called in `SetupFreeListings`. 

However, would be nice to add a JSDOC in this function explaining whats going on.

eason9487 replied:
Added in https://github.com/woocommerce/google-listings-and-ads/pull/1616/commits/d8ecc0824ce15ca8cbfcae5e49c0896ceb517f1b.

puntope replied:
Now is perfect! thanks

puntope commented on js/src/setup-mc/setup-stepper/saved-setup-stepper.js:
```diff
@@ -58,6 +100,15 @@ const SavedSetupStepper = ( { savedStep, onRefetchSavedStep = () => {} } ) => {
+	const initShippingRates = hasResolvedShippingRates ? shippingRates : null;
+	const initShippingTimes = hasResolvedShippingTimes ? shippingTimes : null;
+	const initTargetAudience = targetAudience?.location ? targetAudience : null;
+	const initSettings = settings?.shipping_rate ? settings : null;
```
❓ 💅  I did some degugging. Settings is `null` by default until populate the object values... so maybe we can just pass `settings` directly ????

puntope replied:
Same for `initTargetAudience ` I guess

eason9487 replied:
It's waiting for completing the above processing to auto-save the default values for `settings`. Passing `settings` directly will lead to the `<Form>` receiving `initialValues` without those default values of `settings` when the `GET mc/settings` request is the last one to respond.

The problem can be simulated by adding delay to [fetchSettings](https://github.com/woocommerce/google-listings-and-ads/blob/43a2f76deda2218b5943be595bb6a1d0b1c5ee47/js/src/data/actions.js#L290):

```js
import { __unstableAwaitPromise } from '@wordpress/data-controls';

export function* fetchSettings() {
	yield __unstableAwaitPromise(
		new Promise( ( resolve ) => setTimeout( resolve, 3000 ) )
	);

	// ... omitted
}
```

using `settings` directly:

```js
<SetupFreeListings
	targetAudience={ initTargetAudience }
	settings={ settings }
	// ... omitted
>
```

deleting saved settings:

```sql
DELETE FROM wp_options WHERE option_name = 'gla_merchant_center';
```

and then refresh page to view the initial shipping options in step 2:

![2022-08-12 12 55 11](https://user-images.githubusercontent.com/17420811/184287554-004d2b2d-c31e-4a66-a31b-ff4a289bd6ad.png)

puntope replied:
Gotcha thanks.

puntope reviewed (commented):
- I checked the behaviour. Tested locally and I didn't find any error. All the combos worked fine for me. 
- I checked also the code. Since there is a lot of files deleted and moved around mixed with changes in those files. It was a bit hard to differentiate what was added/changed and what was just a copy paste for the deleted files. However everything looks good for me... left a few comments and minor questions

puntope commented:
> Not sure why but mc/target_audience will respond with an 400 error if saving with empty countries. From my memory, it should be ok before because the UI allows merchants to unselect all. 🤔

I was unable to save with empty value in the audience selector. 🤔 I guess do you mean it shows an error when the autosave runs when the selector is empty?

eason9487 committed:
Add JSDoc for the `handleFormChange` function in SavedSetupStepper component.

Address https://github.com/woocommerce/google-listings-and-ads/pull/1616#discussion_r943456845

eason9487 commented:
Thanks for the review, @puntope

> * I checked also the code. Since there is a lot of files deleted and moved around mixed with changes in those files. It was a bit hard to differentiate what was added/changed and what was just a copy paste for the deleted files.

Sorry for the inconvenience, I was trying to separate this as smaller PRs but the second one would only be including file deletions. So I ended up choosing to use deal with it altogether.

There are no deleted file changes that were caused by copy-and-paste. All deleted files were actually deleted (20e758fda9ddbfb22e5d2037177e9d69deaad056). And two deleted-and-added files are completely different:
- `js/src/components/free-listings/choose-audience-section/index.js`
- `js/src/components/free-listings/choose-audience-section/choose-audience-section.scss`

> I guess do you mean it shows an error when the autosave runs when the selector is empty?

Yes. UI allows the operation to clear the value in the audience selector but then it causes an error. It may confuse merchants. I will check it later while working on this project and open a follow-up issue if needed.

puntope commented:
> Yes. UI allows the operation to clear the value in the audience selector but then it causes an error. It may confuse merchants. I will check it later while working on this project and open a follow-up issue if needed.

I agree
